### DISCUSSION TITLE
How to run local tracker & peers for development

### DISCUSSION THREAD

**Discussion by dradetsky** [transmission/transmission#5130] [Q&A]
I'm trying to do a little transmission development. I may or may not actually end up finishing the thing I'm working on, but...

When developing a feature, I'd like to be able to test it for an actual download. For various reasons, I'd prefer to be able to do this in a controlled & local way. So what I'd like is some simple way to easily start up a tiny tracker and one or more peers on my local machine which are seeding one or more small reference torrents. Does such a thing exist?

Ideally, I'd want something like a helm chart that I could shove into a local minikube, or something along those lines, with options to configure the number of peers and which reference torrents they're providing and that kind of thing. Then I could feed the (possibly generated for my configuration) torrent into my client build, and it would connect to my local test tracker, locate my local test peers, and try to download from them. Or, more likely, exit with a SIGSEGV because I'm a terrible C++ programmer, but still.

Does anything like this exist? Or is there some other way that developers fake a torrent environment for dev purposes?

**Comment by ckerr** [comment]
We maintainers don't have anything like that set up but I'd love to have it.

Usually I run ad-hoc peers and webseeds on my local network, but that's not as repeatable or amenable to CI, so it's not ideal.

If you decide to make a Swarm As A Service please LMK :smiley_cat:

**Comment by dradetsky** [comment]
I wasn't really thinking about something for CI, but just for exploratory development. Just to exercise my feature in the simplest way possible and see what happens.

I mean, ideally I'd understand the code well enough that I'd be able to prove the feature was probably implemented correctly just using unit tests. But especially when you're trying to jump into a codebase you don't really know, exploratory development can be more effective, or at least I find it is for myself.

So if I'm following you correctly, you basically just have some scripts you use for yourself to start a tracker and peers (I mean, I assume you run a tracker), and there's nothing shared or standard for developers. Correct?

Anyway, it's unlikely that I'm going to roll some kind of portable local swarm setup, but if I did: I strongly doubt that the same tool could be used both to facilitate exploratory development AND be used in CI, especially if you're using GHA (which is very limited; just look at [this](https://docs.github.com/en/actions/using-containerized-services/about-service-containers) if you're interested). I mean, you could, but the resulting tool would be awful for both uses.
